### ISSUE TITLE
proposal: prefer_shape_decoration

### ISSUE THREAD

**Issue by bernaferrari** [dart-lang/linter#3411]
## Description

Flutter has Container, Container has decoration, most people (including me) use BoxDecoration because that's how we learned it and how it is in 99% of tutorials. Most other widgets, however, expect a `shape: OutlinedBorder`, like `RoundedRectangleBorder`, which BoxDecoration doesn't support, but ShapeDecoration does. Instead of having to learn how to use BoxDecoration and shapes, people could learn only one and always rely on shapes. To help with this, we could have a lint that helps them remember about it and keep the code-base with the same objects, easier to reuse and understand.

## Kind

This helps the whole codebase use the same code, which is great for reusability. I wish I could rewrite my older apps to use RoundedRectangleDecoration everywhere (maybe put in a variable and reuse between Containers and Materials).

## Good Examples 

```dart
Container(
  decoration: ShapeDecoration(
    shape: RoundedRectangleBorder(
      borderRadius: BorderRadius.circular(20),
      side: Border.all(),
    ),
  ),
)

Container(
  decoration: ShapeDecoration(
    shape: CircleBorder(),
  ),
)
```

## Bad Examples

```dart
Container(
  decoration: BoxDecoration(
    border: Border.all(),
    borderRadius: BorderRadius.circular(20),
  ),
)

Container(
  decoration: BoxDecoration(
    shape: BoxShape.circle,
  ),
)
```

In my real life, every month or so I type `shape: RoundedBorderDecoration()` in `BoxDecoration`, which is invalid. A lint like this could be precious to remember there is another place for RoundedBorderDecoration.

**Comment by jacob314** [comment]
Can `BoxDecoration` be marked as deprecated? If so, that would be a natural way to possibly wire this up using data driven fixes.

**Comment by bernaferrari** [comment]
> Can BoxDecoration be marked as deprecated?

I think this would be awesome (and radical), but I'm not sure how feasible yet. Documentation states BoxDecoration is "optimized for rectangles specifically". Maybe ShapeDecoration also got optimized during the years, maybe it is really slower. If we had a lint + a quick fix, applying on flutter codebase and waiting for the benchmarks would be nice 😛.

**Comment by goderbauer** [comment]
I don't think we should deprecate BoxDecoration. It is just a more specific Decoration class. Yes, you can do most (all?) of what it implements with ShapeDecoration, but its implementation and API is slightly more complicated. 

If people want this lint, I am not opposed to having it. But I also wouldn't add it to the flutter recommended set of lints...

/cc @Hixie who may have more historic context on why these two are separate.

**Comment by jacob314** [comment]
Given this, I don't think we should have a lint to prefer `ShapeDecoration` and perhaps we should actually have a lint+quickfix to prefer `BoxDecoration` for cases where a `BoxDecoration` works. What we could consider is a quick assist to convert a `BoxDecoration` to a `ShapeDecoration` that could be surfaced as a quickfix any time there is an error due to code requiring a ShapeDecoration rather than a BoxDecoration.

A somewhat similar case is `StatelessWidget` and `StatefulWidget`. Suggesting using `ShapeDecoration` everywhere seems a bit like suggesting using `StatefulWidget` everywhere just in case you need to add some mutable state to a widget later. For that case, we recommend using `StatelessWidget` where possible and provide assists to make it easy to upgrade a `StatelessWidget` to a `StatefulWidget`

**Comment by bernaferrari** [comment]
ShapeDecoration helps because you can reuse the same code everywhere. Stateless and Stateful is just a `widget.` difference, it is not significant. You can transform any `RoundedRectangleBorder` into `BoxDecoration` (maybe prefer_box_decoration, another proposal, the opposite of prefer_shape_decoration lol), it always works, but it is not practical for me because then I have 2 different apis to learn and use.

**Comment by jacob314** [comment]
What I'm suggesting is we could add a quick-assist that converts a `BoxDecoration` into the equivalent `OutlinedBorder`. Would that solve the problem of making you rewrite half the method? Now it should be just one click to to go back and forth between `BoxDecoration` and `OutlinedBorder` or vise versa. 

To be clear, the equivalent case with widgets is quite painful without assists as well. If you have a `StatelessWidget` and need to add state now you need to refactor your `build` method adding `widget.` everywhere to account for the fact that the Build method is now in the `State` object not the `Widget` object.

**Comment by bernaferrari** [comment]
Yes, I think it is fine having two methods, but I see more of personal taste, like `prefer_` lints, not something that will be enforced.

**Comment by jacob314** [comment]
I agree that this in the subset of lints that we couldn't enable by default.

**Comment by Hixie** [comment]
BoxDecoration supports things ShapeDecoration doesn't, like per-side colors. IIRC BoxDecoration is also slightly more efficient? My information may be out of date. There's probably also differences in how it animates to/from other decorations.

**Comment by Hixie** [comment]
(Specifically, regarding the borders, IIRC ShapeDecoration uses the ShapeBorder.paint API, while BoxDecoration uses the BoxBorder.paint API, which means it can do per-side colors with rounded corners.)

**Comment by srawlins** [comment]
I don't think we can support implementing super specific guidelines like this in the linter package. If mis-use of ShapeDecoration and BoxDecoration become such a problem that it makes it into a Flutter Style Guide, I can see implementing something. Otherwise I don't think this is high enough value to implement.
